gh-127443: Update refcounts data with stable ABI functions.

picnixz opened the pull request:
Since the changes were done manually and are mechanical, I'm pretty sure there are some +1 missing or incorrectly deduced. I've added some comments here and there about questions I had. 

<!-- gh-issue-number: gh-127443 -->
* Issue: gh-127443
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--127444.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

picnixz committed:
Update refcounts data with stable ABI functions.

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -277,6 +340,9 @@ PyCode_NewEmpty:int:firstlineno::
 PyCodec_Register:PyObject*:search_function:+1:
 
+PyCodec_Unregister:int:::
+PyCodec_Unregister:PyObject *:search_function:0:
```
Should not it be -1?

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -525,6 +594,11 @@ PyDict_GetItem:PyObject*::0:
+PyDict_GetItemRef:int:::
+PyDict_GetItemRef:PyObject *:p:0:
+PyDict_GetItemRef:PyObject *:key:0:
+PyDict_GetItemRef:PyObject **:result:+1:
```
I am not sure how does it work with double pointers. Are there precedences?

BTW, are there any recirds with non-empty 5th field, and if there are any, what is its meaning?

picnixz replied:
> I am not sure how does it work with double pointers. Are there precedences?

One precedent is 

```text
PyErr_GetExcInfo:void:::
PyErr_GetExcInfo:PyObject**:ptype:+1:
PyErr_GetExcInfo:PyObject**:pvalue:+1:
PyErr_GetExcInfo:PyObject**:ptraceback:+1:
```

I think the convention here is that we set the destination pointer to a new reference of the item we found, effectively increasing the reference count of the underlying object.

picnixz replied:
> BTW, are there any recirds with non-empty 5th field, and if there are any, what is its meaning?

Yes, the meaning is just a "comment".

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -1399,6 +1627,11 @@ PyMethod_New:PyObject*:class:0:
 
+PyModule_Add:int:::
+PyModule_Add:PyObject *:module:0:
+PyModule_Add:const char *:name:0:
```
Refcount is not needed here.

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -1399,6 +1627,11 @@ PyMethod_New:PyObject*:class:0:
+PyModule_Add:int:::
+PyModule_Add:PyObject *:module:0:
+PyModule_Add:const char *:name:0:
+PyModule_Add:PyObject *:value:0:stolen
```
Should not the refcount be -1?

picnixz replied:
There is a comment at the top of the file that says:

> XXX NOTE: the 0/+1/-1 refcount information for arguments is
confusing!  Much more useful would be to indicate whether the
function "steals" a reference to the argument or not.  Take for
example PyList_SetItem(list, i, item).  This lists as a 0 change for
both the list and the item arguments.  However, in fact it steals a
reference to the item argument!

Technically, the reference count of the value itself does not change even though you're stealing the reference.

Eclips4 replied:
Tangentially related: can we add a `stolen` comment to the `PyList_SetItem`? Though the comment at the beginning of the file explicitly says this, I still think it would be useful to add it.

picnixz replied:
I think it would be good but I'd like to address the stealing references separately (easier for reviewers I think).

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -1769,14 +2099,26 @@ PyObject_GetIter:PyObject*:o:0:
 PyObject_HasAttrString:PyObject*:o:0:
 PyObject_HasAttrString:const char*:attr_name::
 
+PyObject_HasAttrWithError:int:::
```
Duplicated.

picnixz replied:
Thanks! it should have been PyObject_HasAttrStringWithError

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -2295,6 +2745,13 @@ PyTime_FromTimeAndFold:int:second::
+PyTraceBack_Here:PyFrameObject *:frame:+1:
+
+PyTraceBack_Print:int:::
+PyTraceBack_Print:PyObject *:traceback:+1:
```
Should not the refcount be 0?

picnixz replied:
Yes, I wrongly C/C the previous one :)

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -2295,6 +2745,13 @@ PyTime_FromTimeAndFold:int:second::
+
+PyTraceBack_Print:int:::
+PyTraceBack_Print:PyObject *:traceback:+1:
+PyTraceBack_Print:PyObject *:filename:+1:
```
Should not the refcount be 0?

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -2406,6 +2899,24 @@ PyType_Ready:PyTypeObject*:type:0:
 PyUnicode_1BYTE_DATA:PyObject*:o:0:
 
+PyUnicode_Append:void:::
+PyUnicode_Append:PyObject **:dst:+1:
```
Should not the refcount be 0?

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -2406,6 +2899,24 @@ PyType_Ready:PyTypeObject*:type:0:
+PyUnicode_Append:PyObject *:src:0:
+
+PyUnicode_AppendAndDel:void:::
+PyUnicode_AppendAndDel:PyObject **:dst:+1:
```
Should not the refcount be 0?

picnixz replied:
Yes. My bad! And PyUnicode_Append also was marked with dst:+1 but it should be 0.

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -2483,6 +2994,26 @@ PyUnicode_FromEncodedObject:PyObject*:obj:0:
 PyUnicode_FromEncodedObject:const char*:encoding::
 PyUnicode_FromEncodedObject:const char*:errors::
 
+PyUnicode_AsDecodedObject:PyObject *:::deprecated
```
Should not the refcount be +1?

picnixz replied:
Yes, I forgot to put it!

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -2863,6 +3432,61 @@ PyUnicodeDecodeError_SetStart:int:::
+PyUnicodeEncodeError_GetEncoding:PyObject *::+1:
+PyUnicodeEncodeError_GetEncoding:PyObject *:exc:0:
+
+PyUnicodeEncodeError_GetEnd:Py_ssize_t:::
```
Is it the right signature? It looks weird.

picnixz replied:
No, it's wrong but I C/C from the existing PyUnicodeDecodeError which had the incorrect signatures. I'm also fixing them at the same time.

serhiy-storchaka commented on Doc/data/refcounts.dat:
```diff
@@ -2969,20 +3627,56 @@ Py_GetProgramFullPath:wchar_t*:::
+Py_IncRef:void:::
+Py_IncRef:PyObject *:o:+1:
+
 Py_NewRef:void:::
```
Is it the correct return type?

picnixz replied:
Obviously not, but I missed it. Good catch.

serhiy-storchaka reviewed (commented):
Thank you for your work. I left some questions, but more thoughtful review I can only do in a week.

picnixz committed:
address Serhiy's review

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -651,6 +744,10 @@ PyErr_Print:void:::
 PyErr_PrintEx:void:::
 PyErr_PrintEx:int:set_sys_last_vars::
 
+PyErr_ProgramText:PyObject *:+1::
```
Reference count effect is the wrong spot
```suggestion
PyErr_ProgramText:PyObject *::+1:
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -1118,6 +1255,11 @@ PyInterpreterState_Clear:PyInterpreterState*:interp::
 
+PyInterpreterState_Get:PyInterpreterState *:::
+
+PyInterpreterState_GetDict:PyObject *:+1::
```
Same here:
```suggestion
PyInterpreterState_GetDict:PyObject *::+1:
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -2951,12 +3597,24 @@ Py_FdIsInteractive:const char*:filename::
 
 Py_GetCopyright:const char*:::
 
+Py_GetConstant:PyObject *:+1:::
```
Once again. Also an extra : at the end
```suggestion
Py_GetConstant:PyObject *::+1:
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -2951,12 +3597,24 @@ Py_FdIsInteractive:const char*:filename::
+Py_GetConstant:PyObject *:+1:::
+Py_GetConstant:unsigned int:constant_id::
+
+Py_GetConstantBorrowed:PyObject *:0:::
```
```suggestion
Py_GetConstantBorrowed:PyObject *::0:
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -2969,20 +3627,56 @@ Py_GetProgramFullPath:wchar_t*:::
+Py_IncRef:void:::
+Py_IncRef:PyObject *:o:+1:
+
+Py_NewRef:PyObject *:o:0:
```
This shouldn't have a parameter name:
```suggestion
Py_NewRef:PyObject *::0:
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -37,6 +37,45 @@ PyAnySet_Check:PyObject*:p:0:
+
+PyArg_UnpackTuple:int:::
+PyArg_UnpackTuple:PyObject *:args:0:
+PyArg_UnpackTuple:const char *:name:0:
```
No refcnt affect needed here
```suggestion
PyArg_UnpackTuple:const char *:name::
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -37,6 +37,45 @@ PyAnySet_Check:PyObject*:p:0:
+
+PyArg_VaParse:int:::
+PyArg_VaParse:PyObject *:args:0:
+PyArg_VaParse:const char *:format:0:
```
Not here either
```suggestion
PyArg_VaParse:const char *:format::
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -37,6 +37,45 @@ PyAnySet_Check:PyObject*:p:0:
+PyArg_VaParseTupleAndKeywords:PyObject *:args:0:
+PyArg_VaParseTupleAndKeywords:PyObject *:kw:0:
+PyArg_VaParseTupleAndKeywords:const char *:format:0:
+PyArg_VaParseTupleAndKeywords:const char *:keywords:0:
```
```suggestion
PyArg_VaParseTupleAndKeywords:const char *:format::
PyArg_VaParseTupleAndKeywords:const char *:keywords::
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -1312,6 +1487,16 @@ PyMapping_GetItemString:PyObject*::+1:
+
 PyMapping_HasKey:int:::
 PyMapping_HasKey:PyObject*:o:0:
 PyMapping_HasKey:PyObject*:key::
```
This needs an effect
```suggestion
PyMapping_HasKey:PyObject*:key:0:
```

picnixz replied:
Addressed in https://github.com/python/cpython/pull/127451.

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -1769,14 +2099,26 @@ PyObject_GetIter:PyObject*:o:0:
 
+PyObject_HasAttrStringWithError:int:::
+PyObject_HasAttrStringWithError:PyObject *:o:0:
+PyObject_HasAttrStringWithError:const char *:attr_name:0:
```
No refcnt effect:
```suggestion
PyObject_HasAttrStringWithError:const char *:attr_name::
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -1878,6 +2234,26 @@ PyObject_TypeCheck:int:::
 
+PyObject_Vectorcall:PyObject *::+1:
+PyObject_Vectorcall:PyObject *:callable:0:
+PyObject_Vectorcall:PyObject *const *:args:0:
```
We don't have a reference count effect for any of the other `PyObject *const *` functions. Let's make a decision on whether to have it or not for these.
```suggestion
PyObject_Vectorcall:PyObject *const *:args::
```

picnixz replied:
I prefer to have them. We can think of args as being loosely typed as `PyObject **` and since we are also managing double pointers like that, I think we can also manage those.

ZeroIntensity replied:
Could you fix the other cases then?

picnixz replied:
I'd like to do it in a separate PR. I don't want to change existing lines in this specific PR.

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -1878,6 +2234,26 @@ PyObject_TypeCheck:int:::
+
+PyObject_VectorcallMethod:PyObject *::+1:
+PyObject_VectorcallMethod:PyObject *:name:0:
+PyObject_VectorcallMethod:PyObject *const *:args:0:
```
Same here
```suggestion
PyObject_VectorcallMethod:PyObject *const *:args::
```

picnixz replied:
See https://github.com/python/cpython/pull/127444#discussion_r1864771907.

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -2255,6 +2652,59 @@ PyThreadState_SetAsyncExc:PyObject*:exc:+1:
+PyThread_free_lock:void:::
+PyThread_free_lock:PyThread_type_lock:lock::
+
+PyThread_GetInfo:PyObject *:::
```
Needs an effect. It returns a borrowed reference, I think?
```suggestion
PyThread_GetInfo:PyObject *::0:
```

picnixz replied:
No, a strong one (it constructs a new structure object):

```C
    threadinfo = PyStructSequence_New(&ThreadInfoType);
```

ZeroIntensity commented on Doc/data/refcounts.dat:
```diff
@@ -2863,6 +3432,61 @@ PyUnicodeDecodeError_SetStart:int:::
+PyUnicodeTranslateError_SetStart:Py_ssize_t:start::
+
 PyWeakref_Check:int:::
 PyWeakref_Check:PyObject*:ob::
```
This, as well as `PyWeakref_CheckRef` and `PyWeakref_CheckProxy` below it, need reference count effects.
```suggestion
PyWeakref_Check:PyObject*:ob:0:
```

picnixz replied:
Addressed in https://github.com/python/cpython/pull/127451.

ZeroIntensity reviewed (commented):
Some other things that I found that you didn't modify:

- `PyErr_NormalizeException` is missing a reference count effect for a parameter.
- So is `PySeqIter_New`.
- `PySequence_FAST_ITEMS` is missing an effect for its return value (double check me on that, it returns `PyObject **`).
- `PyUnicode_AsUTF8AndSize` erroneously has a reference count effect for a `Py_ssize_t *` parameter. 
- Some functions that take `PyModuleDef` are missing reference count effects for the passed module def. Namely, `PyModule_Create`, `PyModule_Create2`, `PyModule_ExecDef`, `PyModule_FromDefAndSpec`, `PyModule_FromDefAndSpec2`, `PyState_AddModule`, `PyState_FindModule`, `PyState_RemoveModule`, and `PyType_GetModuleByDef`.

picnixz committed:
address Peter's review

picnixz committed:
Address Peter's review (round 2)

picnixz committed:
add missing stable ABI entries

picnixz committed:
complete TODOs

picnixz committed:
Resolve some TODOs.

In PyErr_SetHandledException, the exception's reference count is incremented by 1 since it's being used as `Py_XSETREF(..., Py_NewRef(exc))`.

In PyType_FromModuleAndSpec, the module's reference count is incremented by 1 since it's being stored in the new class being created.

picnixz committed:
Merge remote-tracking branch 'upstream/main' into doc/refcounts/update-127443

encukou commented on Doc/data/refcounts.dat:
```diff
@@ -1641,9 +1930,63 @@ PyOS_AfterFork_Parent:void:::
+PyOS_string_to_double:double:::
+PyOS_string_to_double:const char *:s::
+PyOS_string_to_double:char **:endptr::
+PyOS_string_to_double:PyObject *:overflow_exception:+1:
```
Is +1 correct here? It's ±0 on success.

encukou commented on Doc/data/refcounts.dat:
```diff
@@ -3016,10 +3812,50 @@ _PyBytes_Resize:int:::
 _PyBytes_Resize:PyObject**:bytes:0:
 _PyBytes_Resize:Py_ssize_t:newsize::
 
+_PyObject_CallFunction_SizeT:PyObject *::+1:abi-only
```
Please pick only one of `abi_only` or `abi-only`.
